ecs-cli/modules/cli/regcreds/create_task_execution_role.go

```diff
@@ -0,0 +1,149 @@
+	policyDescriptionFmtString = "Policy generated by the ecs-cli for role %s"
+	roleDescriptionString      = "Role generated by the ecs-cli"
+)
+
+type executionRoleParams struct {
+	CredEntries map[string]CredsOutputEntry
+	RoleName    string
+	Region      string
+	IAMClient   iamClient.Client
+	KMSClient   kmsClient.Client
+}
+
+func createTaskExecutionRole(params executionRoleParams) error {
+	log.Infof("Creating resources for task execution role %s...", params.RoleName)
+
+	// generate policy document
+	policyDoc, err := generateSecretsPolicy(params.CredEntries, params.KMSClient)
+	if err != nil {
+		return err
+	}
+
+	// create the new policy
+	newPolicy, err := createRegistryCredentialsPolicy(params.RoleName, policyDoc, params.IAMClient)
+	if err != nil {
+		return err
+	}
+	newPolicyArn := *newPolicy.Arn
+	log.Infof("Created new task execution role policy %s", newPolicyArn)
+
+	// create role
+	roleName, err := createOrFindRole(params.RoleName, params.IAMClient)
+	if err != nil {
+		return err
+	}
+
+	// attach managed execution role policy & new credentials policy to role
+	err = attachRolePolicies(newPolicyArn, roleName, params.Region, params.IAMClient)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createRegistryCredentialsPolicy(roleName, policyDoc string, client iamClient.Client) (iam.Policy, error) {
+	newPolicyName := generateECSResourceName(roleName + "-policy")
+	createPolicyRequest := iam.CreatePolicyInput{
+		PolicyName:     newPolicyName,
+		PolicyDocument: aws.String(policyDoc),
+		Description:    aws.String(roleDescriptionString),
+	}
+
+	policyResult, err := client.CreatePolicy(createPolicyRequest)
+	if err != nil {
+		// if policy already exists, create new one with timestamp
+		// TODO: consider instead updating existing policy with CreatePolicyVersion()
+		if entityAlreadyExists(err) {
+			createPolicyRequest := iam.CreatePolicyInput{
+				PolicyName:     generateECSResourceName(roleName + "-policy-" + time.Now().UTC().Format("20060102150405")),
+				PolicyDocument: aws.String(policyDoc),
+				Description:    aws.String(roleDescriptionString),
+			}
+			policyResult2, err := client.CreatePolicy(createPolicyRequest)
+			if err != nil {
+				return iam.Policy{}, err
+			}
+			return *policyResult2.Policy, nil
+		}
+		return iam.Policy{}, err
+	}
+	return *policyResult.Policy, nil
+}
+
+func createOrFindRole(roleName string, client iamClient.Client) (string, error) {
+	createRoleRequest := iam.CreateRoleInput{
+		AssumeRolePolicyDocument: aws.String(assumeRolePolicyDocString),
+		Description:              aws.String(roleDescriptionString),
+		RoleName:                 aws.String(roleName),
+	}
+	roleResult, err := client.CreateRole(createRoleRequest)
+	// if err is b/c role already exists, OK to continue
+	if err != nil && !entityAlreadyExists(err) {
+		return "", err
+	}
+	// TODO: validate AssumeRolePolicyDocument of existing role?
+
+	if roleResult != nil {
```

Comment:
Why are you checking this instead of checking `if err != nil`

Reply:
Because if an error other than "EntityAlreadyExists" was returned by CreateRole, it would have already been returned on line 110. This check is to see whether a new Role object was returned so we can message the customer accordingly (new role ARN or existing role name).